src/init/common.cpp

```diff
@@ -80,26 +79,31 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
     return {};
 }
 
-util::Result<void> SetLoggingCategories(const ArgsManager& args)
+[[nodiscard]] static bool EnableOrDisable(const std::string& opt, const std::string& category)
 {
-    if (args.IsArgSet("-debug")) {
-        // Special-case: if -debug=0/-nodebug is set, turn off debugging messages
-        const std::vector<std::string> categories = args.GetArgs("-debug");
-
-        if (std::none_of(categories.begin(), categories.end(),
-            [](std::string cat){return cat == "0" || cat == "none";})) {
-            for (const auto& cat : categories) {
-                if (!LogInstance().EnableCategory(cat)) {
-                    return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", cat)};
-                }
-            }
-        }
+    if (opt == "-debug") return LogInstance().EnableCategory(category);
+    if (opt == "-debugexclude") return LogInstance().DisableCategory(category);
+    NONFATAL_UNREACHABLE();
+}
+
+// Maintain similar logic in both of the EnableOrDisableLogCategories() functions in the codebase.
+[[nodiscard]] static util::Result<void> EnableOrDisableLogCategories(const ArgsManager& args, const std::string& opt)
+{
+    const std::vector<std::string>& categories{args.GetArgs(opt)};
+    if (std::any_of(categories.cbegin(), categories.cend(), [](const auto& c) { return LogInstance().IsNoneCategory(c); })) return {};
+    for (const auto& c : categories) {
+        if (!EnableOrDisable(opt, c)) return util::Error{strprintf(_("Unsupported logging category %s=%s."), opt, c)};
     }
+    return {};
+}
 
-    // Now remove the logging categories which were explicitly excluded
-    for (const std::string& cat : args.GetArgs("-debugexclude")) {
-        if (!LogInstance().DisableCategory(cat)) {
-            return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debugexclude", cat)};
+util::Result<void> SetLoggingCategories(const ArgsManager& args)
+{
+    // debugexclude settings take priority over debug ones, so run debugexclude last
+    for (const std::string& opt : {"-debug", "-debugexclude"}) {
```

Comment:
This seems a bit clunky, how about:

```c++
// Maintain similar logic in both of the EnableOrDisableLogCategories() functions in the codebase.
static void EnableOrDisableLogCategories(util::Result<void>& result, const ArgsManager& args, const std::string& opt, bool enable)
{
    if (!result) return;
    if (!args.IsArgSet(opt)) return;
    const std::vector<std::string>& categories{args.GetArgs(opt)};
    if (std::any_of(categories.cbegin(), categories.cend(), [](const auto& c) { return LogInstance().IsNoneCategory(c); })) return;
    for (const auto& category : categories) {
        bool success = (enable ? LogInstance().EnableCategory(category) : LogInstance().DisableCategory(category));
        if (!success) {
            result = util::Error{strprintf(_("Unsupported logging category %s=%s."), opt, category)};
            return;
        }
    }
    return;
}

util::Result<void> SetLoggingCategories(const ArgsManager& args)
{
    util::Result<void> result;

    // debugexclude settings take priority over debug ones, so run debugexclude last
    EnableOrDisableLogCategories(result, args, "-debug", /*enable=*/true);
    EnableOrDisableLogCategories(result, args, "-debugexclude", /*enable=*/false);
    if (!result) return util::Error{util::ErrorString(result)};
    return {};
}
```

(or just `return result;` at the end -- not sure what the point of `return Error(ErrorString(result))` is?)

Reply:
Great suggestion AJ this are looks a lot better now